src/execution_plan/execution_plan_clone.c

```diff
@@ -8,92 +8,69 @@
 #include "../query_ctx.h"
 #include "../util/rax_extensions.h"
 
-// Clones an execution plan operations, with respect to the original execution plan segment.
-static OpBase *_ExecutionPlan_CloneOperations(const ExecutionPlan *orig_plan,
-											  ExecutionPlan *clone_plan, const OpBase *op) {
-	// If there is no op, or the op is a part of a different segment, return NULL.
-	if(!op || op->plan != orig_plan) return NULL;
+static ExecutionPlan *_ClonePlanInternals(const ExecutionPlan *template) {
+	ExecutionPlan *clone = ExecutionPlan_NewEmptyExecutionPlan();
 
-	// Clone the op.
-	OpBase *clone = OpBase_Clone(clone_plan, op);
-	// Clone the op's children and add them the cloned children array.
-	for(uint i = 0; i < op->childCount; i++) {
-		OpBase *cloned_child = _ExecutionPlan_CloneOperations(orig_plan, clone_plan,
-															  op->children[i]);
-		// Assumption: all the children are either in the same segment or on a different segment.
-		if(!cloned_child) break;
-		ExecutionPlan_AddOp(clone, cloned_child);
+	clone->record_map = raxClone(template->record_map);
+	if(template->ast_segment) clone->ast_segment = AST_ShallowCopy(template->ast_segment);
+	if(template->query_graph) clone->query_graph = QueryGraph_Clone(template->query_graph);
```

Comment:
I think we can avoid both `query_graph` and `connected_components` if we introduce ref count to the QueryGraph object.